app/models/concerns/foreman_bootdisk/compute_resources/proxmox.rb

```diff
@@ -0,0 +1,50 @@
+# frozen_string_literal: true
+
+module ForemanBootdisk
+  module ComputeResources
+    module Proxmox
+      def capabilities
+        super + [:bootdisk]
+      end
+
+      def iso_upload(iso, vm_uuid)
+        server = find_vm_by_uuid(vm_uuid)
+        server.ssh_options = { password: fog_credentials[:proxmox_password] }
+        server.ssh_ip_address = server.node_id
+        server.username = client.credentials[:current_user].split('@').first
+        server.scp_upload(iso, '/var/lib/vz/template/iso/')
+        server.reload
+        storage = storages(server.node_id, 'iso')[0]
+        storage.volumes.any? { |v| v.volid.include? File.basename(iso) }
+      end
+
+      def iso_attach(iso, vm_uuid)
+        server = find_vm_by_uuid(vm_uuid)
+        storage = storages(server.node_id, 'iso')[0]
+        volume = storage.volumes.detect { |v| v.volid.include? File.basename(iso) }
+        disks = server.disks.map { |disk| disk.split(":")[0] }.join(";")
+        config_hash = { ide2: "#{volume.volid},media=cdrom" }
+        server.update(config_hash)
+        server.update({ boot: "order=ide2;#{disks}" })
+        server.reboot
+      end
+
+      def iso_detach(vm_uuid)
+        server = find_vm_by_uuid(vm_uuid)
+
+        # get volid to delete iso after detaching from vm
+        volid = server.volumes.get("ide2").volid
+        config_hash = { ide2: "none,media=cdrom" }
+        server.update(config_hash)
+
+        # cdrom will be ejected on next power off
+        server.detach('ide2')
+
+        # delete the iso file from proxmox server
+        storage = storages(server.node_id, 'iso')[0]
+        volume = storage.volumes.detect { |v| v.volid.include? volid }
```

Comment:
When would this volume ID not be exactly equal? If the volume IDs are integers I'd imagine this also end up being `123'.include?('12')` so
```suggestion
        volume = storage.volumes.detect { |v| v.volid == volid }
```

Also, previously you already used `volid = server.volumes.get("ide2").volid` so why can't you just do:
```
server.volumes.get('ide2').destroy
```

Reply:
We don't want to destroy the disk 'ide2' as the operation is not possible when vm is powered on. We just want to destroy the volume so that the `iso files` are not piled up in proxmox server.

Reply:
So `volume = storage.volumes.detect { |v| v.volid.include? volid }` returns a different object than `server.volumes.get("ide2")`. That would surprise me.

Reply:
Yes, server.volumes are disks of the server so it returns the disks, `Fog::Proxmox::Compute::Disks`  and storage.volumes are volumes so it returns `Fog::Proxmox::Compute::Volumes`. You can remove the volumes without caring about server but you cannot remove the disk without powering off the server (hotplug issue).